### PR TITLE
Fix replacement refinement with empty tiles

### PR DESCRIPTION
**Pull request by lilleyse**
For #3241 

I tried another approach at solving replacement refinement when empty tiles are involved. `selectTiles` is now back to its original state for the most part. The new idea is to find the descendant tiles as a pre-process in `loadTileset`. Then at the end of `selectTiles` the refining tiles check if all their descendant tiles were actually selected this frame. If not, the refining tile is selected and its descendants are deselected. So it's more of a reactive approach, which is actually safer because I noticed some edge cases with the previous approach.

The main drawback is in order to "deselect" a tile, I added a boolean in `Cesium3DTile` called `selected`. There's probably a better way to do this, but I was worried that modifying the `selectedTiles` array would cause a performance hit.

One final thing, I experimented with removing the `tilesetContent` flag because it doesn't serve much of a purpose anymore, except to speed up traversal slightly for those nodes. It works with or without it, but I kept it in for now.

### CODE REVIEW

**Comment by pjcozzi** [comment]
Thanks @lilleyse, this approach is good with me - and sounds cleaner.

@e-andersson do you want to do an initial review?

**Comment by lilleyse** [comment]
Also just a note for reviewing, it's a lot clearer to look at 3cf51b5 alone.

File: Source/Scene/Cesium3DTileset.js [hunk 1]
```diff
@@ -367,13 +403,14 @@ define([
         // zoomed into a neighborhood and can cull the skyscrapers in the root node.
         //
         // Don't select if the tile is being loaded to refine another tile
```
Comment (e-andersson) [inline Source/Scene/Cesium3DTileset.js]: This comment can be removed now, doesn't apply anymore

File: Source/Scene/Cesium3DTileset.js [hunk 2]
```diff
@@ -309,6 +320,31 @@ define([
         });
     };
 
+    function loadRefiningTiles(refiningTiles) {
```
Comment (e-andersson) [inline Source/Scene/Cesium3DTileset.js]: Would it be a good idea to call this function something else than `load...` to make it sound less like having to do with loading tile content? `prepareRefiningTiles`, `trackDescendantsWithContent` or something along those lines?

**Comment by e-andersson** [comment]
This looks pretty good. No particular comments except for the those two comments above. The approach seems sound, but I haven't tried running it yet. (Will try it, but may not be able to get to it today.)

**Comment by pjcozzi** [comment]
Thanks for the reivew, @e-andersson!

File: Source/Scene/Cesium3DTile.js [hunk 3]
```diff
@@ -120,6 +120,23 @@ define([
...
 
         /**
+         * Descendant tiles that need to be visible before this tile can refine. For example, if
+         * a child is empty (such as for accelerating culling), its descendants with content must
+         * be loaded first. If a tiles's children all have content, this is left undefined.
+         *
+         * @type {Array}
+         * @readonly
+         */
+        this.descendantsWithContent = undefined;
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTile.js]: Perhaps mention in the comment that this is computed at runtime, not loaded from tileset.json.

Also, as we discussed before, but now that you have more context, would it be useful for tileset.json to have any metadata for this, e.g., number of descendants with content?

File: Source/Scene/Cesium3DTile.js [hunk 4]
```diff
@@ -120,6 +120,23 @@ define([
...
+
+        /**
+         * Marks if the tile is selected this frame.
+         *
+         * @type {Boolean}
+         */
+        this.selected = false;
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTile.js]: Move this down to the section
```
// Members that are updated every frame for rendering optimizations:
```
and perhaps update that comment since the dynamic per-frame members are now for more than optimization.

File: Source/Scene/Cesium3DTile.js [hunk 5]
```diff
@@ -389,8 +378,11 @@ define([
      * DOC_TBA
      */
     Cesium3DTile.prototype.update = function(owner, frameState) {
-        applyDebugSettings(this, owner, frameState);
-        this._content.update(owner, frameState);
+        if (this.selected) {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTile.js]: In practice, it will not matter, but if you want to be cool and optimize to avoid a cache miss change this to:
```javascript
        if (this.selected) {
            this.selected = false;
            applyDebugSettings(this, owner, frameState);
            this._content.update(owner, frameState);
        }
```
Otherwise, we risk `this._content.update` kicking `selected` out of a register (almost most certainty) and maybe even out of L2 (if, for example, it called into WebGL).

File: Source/Scene/Cesium3DTileset.js [hunk 6]
```diff
@@ -286,22 +288,31 @@ define([
...
                         stack.push({
                             header : childHeader,
                             cesium3DTile : childTile
                         });
+                        if (!childTile.hasContent) {
+                            hasEmptyChild = true;
+                        }
                     }
                 }
+                if (tile3D.hasContent && hasEmptyChild && (tile3D.refine === Cesium3DTileRefine.REPLACE)) {
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Might want to have a comment here explaining why all this is being checked.

File: Source/Scene/Cesium3DTileset.js (hunk 2 above)
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: +1

File: Source/Scene/Cesium3DTileset.js [hunk 7]
```diff
@@ -309,6 +320,31 @@ define([
         });
     };
 
+    function loadRefiningTiles(refiningTiles) {
+        // Tiles that use replacement refinement and have empty child tiles need to keep track of
+        // descendants with content in order to refine correctly.
+        var stack = [];
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Do you want to generalize the tree traversal, and then pass a function to it here and in `destroy`?

File: Source/Scene/Cesium3DTileset.js [hunk 8]
```diff
@@ -367,13 +403,14 @@ define([
...
-        if (tile.isReady() && !tile.refining &&
-                (fullyVisible || (tile.contentsVisibility(frameState.cullingVolume) !== Intersect.OUTSIDE))) {
+        if (tile.isReady() && (fullyVisible || (tile.contentsVisibility(frameState.cullingVolume) !== Intersect.OUTSIDE))) {
             selectedTiles.push(tile);
+            tile.selected = true;
         }
     }
 
     var scratchStack = [];
+    var refiningTiles = [];
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Prefix with `scratch`.

File: Source/Scene/Cesium3DTileset.js [hunk 9]
```diff
@@ -515,37 +553,69 @@ define([
...
+        var refiningTilesLength = refiningTiles.length;
+        for (var i = 0; i < refiningTilesLength; ++i) {
+            var j;
+            var refinable = true;
+            var refiningTile = refiningTiles[i];
+            var descendantsLength = refiningTile.descendantsWithContent.length;
+            for (j = 0; j < descendantsLength; ++j) {
+                descendant = refiningTile.descendantsWithContent[j];
+                if (!descendant.selected) {
+                    // TODO: also check that its visible
```
Comment (pjcozzi) [inline Source/Scene/Cesium3DTileset.js]: Do you want to do this `TODO` now or later?

File: Specs/Scene/Cesium3DTilesetSpec.js [hunk 10]
```diff
@@ -185,6 +185,7 @@ defineSuite([
 
             expect(tileset._geometricError).toEqual(240.0);
             expect(tileset._root).toBeDefined();
+            expect(tileset._root.descendantsWithContent).toBeUndefined();
```
Comment (pjcozzi) [inline Specs/Scene/Cesium3DTilesetSpec.js]: Do we need any new tests to cover corner cases or new code paths?

**Comment by pjcozzi** [comment]
Do all the tests pass for you in this branch? They don't for me, but the failures are inconsistent. I think it is timing or my machine.

**Comment by pjcozzi** [comment]
That's all my comments.

File: Source/Scene/Cesium3DTile.js (hunk 3 above)
Comment (lilleyse) [inline Source/Scene/Cesium3DTile.js]: I don't think any metadata would help right now, unless it could link to those descendants directly, which doesn't seem possible.

File: Source/Scene/Cesium3DTileset.js (hunk 7 above)
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Maybe not yet. While some code is duplicated, I think its easier to understand this way.

File: Source/Scene/Cesium3DTileset.js (hunk 9 above)
Comment (lilleyse) [inline Source/Scene/Cesium3DTileset.js]: Since a lot of the work is in https://github.com/AnalyticalGraphicsInc/cesium/pull/3449 this is just a reminder for that PR.

File: Specs/Scene/Cesium3DTilesetSpec.js (hunk 10 above)
Comment (lilleyse) [inline Specs/Scene/Cesium3DTilesetSpec.js]: The new code paths should be tested ok by the examples. The coverage is still at 100%.

**Comment by lilleyse** [comment]
> Do all the tests pass for you in this branch? They don't for me, but the failures are inconsistent. I think it is timing or my machine.

There were some failing tests that I just fixed, and one or two tests that don't seem related. The `debugFreezeFrame` test still fails because of the new `selected` bool in `Cesium3DTileset` being set to false automatically. I suppose `Cesium3DTile` could know if `debugFreezeFrame` is true and not set `selected` to false. In general I think the `selected` boolean is not great, and would rather do something else.

**Comment by pjcozzi** [comment]
We should fix these tests before merging.  Perhaps instead of using a `selected` boolean, we can track the last frame number.

**Comment by lilleyse** [comment]
I organized it differently such that `debugFreezeFrame` works correctly. The tests all pass now.

**Comment by pjcozzi** [comment]
JSHint fails:
```
Specs/Scene/Cesium3DTilesetSpec.js
  line 520  col 22  'root' is already defined.
```

**Comment by lilleyse** [comment]
Oops. Fixed that.

**Comment by pjcozzi** [comment]
Looks good.  Will also merge this into the doc branch now.

**Comment by e-andersson** [comment]
@lilleyse Have you tested this in the viewer with some data? It seems to me that we're back at the problem where blank space is shown when the tileset content tile is selected and has not yet loaded the json data.
